controllers/edgeconfig_controller.go

```diff
@@ -0,0 +1,240 @@
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/project-flotta/flotta-operator/api/v1alpha1"
+	"github.com/project-flotta/flotta-operator/internal/common/repository/edgeconfig"
+	"github.com/project-flotta/flotta-operator/internal/common/repository/edgedevice"
+	"github.com/project-flotta/flotta-operator/internal/common/repository/playbookexecution"
+)
+
+// EdgeConfigReconciler reconciles a EdgeConfig object
+type EdgeConfigReconciler struct {
+	client.Client
+	Scheme                      *runtime.Scheme
+	EdgeConfigRepository        edgeconfig.Repository
+	EdgeDeviceRepository        edgedevice.Repository
+	PlaybookExecutionRepository playbookexecution.Repository
+	Concurrency                 uint
+	MaxConcurrentReconciles     int
+	ExecuteConcurrent           func(context.Context, uint, ConcurrentFunc, []v1alpha1.EdgeDevice) []error
+}
+
+var logger = log.FromContext(context.TODO())
+
+//+kubebuilder:rbac:groups=management.project-flotta.io,resources=edgeconfigs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=management.project-flotta.io,resources=edgeconfigs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=management.project-flotta.io,resources=edgeconfigs/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the EdgeConfig object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+func (r *EdgeConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.V(1).Info("EdgeConfig Reconcile", "edgeConfig", req)
+
+	edgeConfig, err := r.EdgeConfigRepository.Read(ctx, req.Name, req.Namespace)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	if edgeConfig.DeletionTimestamp != nil {
+		return ctrl.Result{}, nil
+	}
+
+	edgeDevices, err := r.EdgeDeviceRepository.ListForEdgeConfig(ctx, edgeConfig.Name, edgeConfig.Namespace)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{Requeue: true}, err
+		}
```

Comment:
if no edgedevices should return here, no?

Reply:
Do you mean I should add a `return ctrl.Result{}, err` or  `return ctrl.Result{true}, err`  in line 88 ?

Reply:
Added. 
I also included a len check on the devices slice.